Bearded.Utilities/Noise/NoiseMap.cs

```diff
@@ -0,0 +1,63 @@
+using System;
+
+namespace Bearded.Utilities.Noise
+{
+    public static class NoiseMap
+    {
+        public static INoiseMap FromArray(double[,] array, IInterpolationMethod2 interpolation) =>
+            new ArrayNoiseMap(array, interpolation);
+
+        private sealed class ArrayNoiseMap : INoiseMap
+        {
+            private readonly double[,] array;
+            private readonly IInterpolationMethod2 interpolation;
+            private readonly int width;
+            private readonly int height;
+
+            public ArrayNoiseMap(double[,] array, IInterpolationMethod2 interpolation)
+            {
+                this.array = array;
+                this.interpolation = interpolation;
+                width = array.GetLength(0);
+                height = array.GetLength(1);
+            }
+
+            public double ValueAt(double x, double y)
+            {
+                if (x < 0 || x >= 1)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(x));
+                }
+                if (y < 0 || y >= 1)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(y));
+                }
+
+                // First we transform x and y from [0, 1) x [0, 1) to [0, width) x [0, height) for easier math.
+                x *= width;
+                y *= height;
+
+                // The noise array keeps track of the values at the center of each "pixel", so for the interpolation, we
+                // need to move the array coordinates by (0.5, 0.5), which is the same as moving the x, y by
+                // (-0.5, -0.5).
+                x -= 0.5;
+                y -= 0.5;
+
+                var xBelow = MoreMath.FloorToInt(x);
```

Comment:
This can easily end up as -1 if x is passed in as 0, right? that doesn't seem right?

Reply:
It is, because the centre of the pixel stored at [0, 0] in the array is not actually at texture coordinates (0, 0) so we need to interpolate between the value in [0, 0] and [-1, -1], where the behaviour of [-1, -1] is determined by how the texture wraps.